src/query/storage/cache/utils.go

```diff
@@ -98,22 +98,57 @@ func samplesEqual(a, b []prompb.Sample, end int64, logger *zap.Logger) bool {
 	return true
 }
 
+// Create new result that doesn't include data past end
+func createEndFilteredTimeseries(timeseries []*prompb.TimeSeries, end int64) []*prompb.TimeSeries {
```

Comment:
looks like many copies in this method, would this cause high cpu or memory?

Reply:
Since the check rate is pretty low (0.5% of all requests), I don't think it should lead to high CPU usage or memory usage

Reply:
Just checked with profiling, checking doesn't seem to contribute to much CPU usage